connexion/security.py

```diff
@@ -22,229 +55,269 @@
-        check_oauth_func = self.check_oauth_func(token_info_func, scope_validate_func)
+        check_bearer_func = self.check_bearer_token(token_info_func)
 
         def wrapper(request):
             auth_type, token = self.get_auth_header_value(request)
             if auth_type != "bearer":
                 return self.no_value
-
-            return check_oauth_func(request, token, required_scopes=required_scopes)
+            return check_bearer_func(request, token)
 
         return wrapper
 
-    def verify_basic(self, basic_info_func):
-        check_basic_info_func = self.check_basic_auth(basic_info_func)
-
-        def wrapper(request):
-            auth_type, user_pass = self.get_auth_header_value(request)
-            if auth_type != "basic":
-                return self.no_value
 
-            try:
-                username, password = (
-                    base64.b64decode(user_pass).decode("latin1").split(":", 1)
-                )
-            except Exception:
-                raise OAuthProblem(detail="Invalid authorization header")
+class ApiKeySecurityHandlerFactory(AbstractSecurityHandlerFactory):
+    """
+    Factory for security for API Keys.
+    """
 
-            return check_basic_info_func(request, username, password)
+    def get_fn(self, security_scheme, required_scopes):
+        apikey_info_func = self._resolve_func(security_scheme)
+        if not apikey_info_func:
+            logger.warning("... x-apikeyInfoFunc missing")
+            return None
 
-        return wrapper
+        return self._get_verify_func(
+            apikey_info_func,
+            security_scheme["in"],
+            security_scheme["name"],
+        )
 
-    @staticmethod
-    def get_cookie_value(cookies, name):
+    def _resolve_func(self, security_definition):
         """
-        Called inside security wrapper functions
+        :type security_definition: dict
+        :rtype: function
 
-        Returns cookie value by its name. None if no such value.
-        :param cookies: str: cookies raw data
-        :param name: str: cookies key
+        >>> _resolve_func({'x-apikeyInfoFunc': 'foo.bar'})
+        '<function foo.bar>'
         """
-        cookie_parser = http.cookies.SimpleCookie()
-        cookie_parser.load(str(cookies))
-        try:
-            return cookie_parser[name].value
-        except KeyError:
-            return None
+        return self._get_function(
+            security_definition, "x-apikeyInfoFunc", "APIKEYINFO_FUNC"
+        )
 
-    def verify_api_key(self, api_key_info_func, loc, name):
+    # TODO: Move logic from here to `get_fn`?
+    #   that has the entire security definition, so should determine which function?
+    # All API key checking funcs have signature func(api_key, required_scopes) -> result
+    def _get_verify_func(self, api_key_info_func, loc, name):
         check_api_key_func = self.check_api_key(api_key_info_func)
 
         def wrapper(request):
```

Comment:
Can't add a comment on the specific line below, but the API key used to be removed from the query string before passing it to the view function. I think this is necessary to prevent an unexpected keyword argument error. If we want to achieve the same behaviour, we should ensure it is removed from the scope.

Reply:
Very good remark, didn't think of that. I suspect this mean that we don't have a test for this now then, otherwise this test should be failing, no?

Reply:
I assume a test is missing indeed. Something to check before merging the PR.

Currently the scope is not available in the `SecurityHandler`s though, which complicates modifying the scope. We could solve this by making the `SecurityHandler`s ASGI callables, which could have some additional nice benefits. 
- It would fit with the "everything is an ASGI callable" design
- It might remove some of the functional complexity we currently have, as the `SecurityHandler` would no longer need to return a function, but can just be called itself